packages/flutter/lib/src/widgets/app_lifecycle_listener.dart

```diff
@@ -0,0 +1,252 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+
+import 'binding.dart';
+
+/// A callback type that is used by [AppLifecycleListener.onExitRequested] to
+/// ask the application if it wants to cancel application termination or not.
+typedef AppExitRequestCallback = Future<AppExitResponse> Function();
+
+/// A listener that can be used to listen to changes in the application
+/// lifecycle.
+///
+/// To listen for requests for the application to exit, and to decide whether or
+/// not the application should exit when requested, create an
+/// [AppLifecycleListener] and set the [onExitRequested] callback.
+///
+/// To listen for changes in the application lifecycle state, define an
+/// [onStateChange] callback. See the [AppLifecycleState] enum for details on
+/// the various states.
+///
+/// The [onStateChange] callback is called for each state change, and the
+/// individual state transitions ([onResume], [onInactive], etc.) are also
+/// called if the state transition they represent occurs. Both [onStateChange]
+/// and the specific state transition callback will be called when the state
+/// changes.
+///
+/// {@tool dartpad} This example shows how an application can listen to changes
+/// in the application state.
+///
+/// ** See code in examples/api/lib/widgets/app_lifecycle_listener/app_lifecycle_listener.0.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad} This example shows how an application can optionally decide
+/// to abort a request for exiting instead of obeying the request.
+///
+/// ** See code in examples/api/lib/widgets/app_lifecycle_listener/app_lifecycle_listener.1.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+/// * [ServicesBinding.exitApplication] for a function to call that will request
+///   that the application exits.
+/// * [WidgetsBindingObserver.didRequestAppExit] for the handler which this
+///   class uses to receive exit requests.
+/// * [WidgetsBindingObserver.didChangeAppLifecycleState] for the handler which
+///   this class uses to receive lifecycle state changes.
+class AppLifecycleListener with WidgetsBindingObserver, Diagnosticable {
+  /// Creates an [AppLifecycleListener].
+  AppLifecycleListener({
+    WidgetsBinding? binding,
```

Comment:
Okay,  This was a suggestion that @Hixie had early on. It does seem convenient, and in normal usage it doesn't make any difference. It would be less complex to not have it, though.